### PR TITLE
provider/openstack: store advanced configs in stack, and return them …

### PR DESCRIPTION
**Pull request by drmaas**
…from cluster/server group provider

### CODE REVIEW

**Comment by drmaas** [comment]
@derekolk @varikin @spinnaker/reviewers

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/description/servergroup/ServerGroupParameters.groovy [hunk 1]
```diff
@@ -47,32 +46,36 @@ class ServerGroupParameters {
...
+      scaleup_cooldown     : scaleup?.cooldown ? scaleup.cooldown.toString() : null,
+      scaleup_adjustment   : scaleup?.adjustment ? scaleup.adjustment.toString() : null,
+      scaleup_period       : scaleup?.period ? scaleup.period.toString() : null,
+      scaleup_threshold    : scaleup?.threshold ? scaleup.threshold.toString() : null,
+      scaledown_cooldown   : scaledown?.cooldown ? scaledown.cooldown.toString() : null,
+      scaledown_adjustment : scaledown?.adjustment ? scaledown.adjustment.toString() : null,
+      scaledown_period     : scaledown?.period ? scaledown.period.toString() : null,
+      scaledown_threshold  : scaledown?.threshold ? scaledown.threshold.toString() : null,
+      source_user_data_type: sourceUserDataType ?: null,
+      source_user_data     : sourceUserData ?: null,
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/description/servergroup/ServerGroupParameters.groovy]: Whats the difference between `source_user_data` and `user_data`?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/description/servergroup/ServerGroupParameters.groovy [hunk 2]
```diff
@@ -47,32 +46,36 @@ class ServerGroupParameters {
...
+      security_groups      : securityGroups ? securityGroups.join(',') : null,
+      autoscaling_type     : autoscalingType ? autoscalingType.toString() : null,
+      scaleup_cooldown     : scaleup?.cooldown ? scaleup.cooldown.toString() : null,
+      scaleup_adjustment   : scaleup?.adjustment ? scaleup.adjustment.toString() : null,
+      scaleup_period       : scaleup?.period ? scaleup.period.toString() : null,
+      scaleup_threshold    : scaleup?.threshold ? scaleup.threshold.toString() : null,
+      scaledown_cooldown   : scaledown?.cooldown ? scaledown.cooldown.toString() : null,
+      scaledown_adjustment : scaledown?.adjustment ? scaledown.adjustment.toString() : null,
+      scaledown_period     : scaledown?.period ? scaledown.period.toString() : null,
+      scaledown_threshold  : scaledown?.threshold ? scaledown.threshold.toString() : null,
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/description/servergroup/ServerGroupParameters.groovy]: I know this isn't new in this change, but couldn't all these ternaries be in the form `scaledown?.threshold?.toString()` and accomplish the same? Same with the `loadBalancers?.join(',')` above?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackServerGroupCachingAgent.groovy [hunk 3]
```diff
@@ -200,6 +200,7 @@ class OpenstackServerGroupCachingAgent extends AbstractOpenstackCachingAgent imp
   OpenstackServerGroup buildServerGroup(ProviderCache providerCache, Stack stack, Set<String> loadbalancerIds) {
     Map<String, Object> launchConfig = buildLaunchConfig(stack.parameters)
     Map<String, Object> openstackImage = buildImage(providerCache, (String) launchConfig?.image)
+    Map<String, Object> advancedConfig = buildAdvancedConfig(stack.parameters)
 
     //TODO this check will change once we have a config that indicates the openstack version, e.g. kilo, liberty, mitaka
     //kilo stack create times have a 'Z' on the end. It was removed in liberty and mitaka.
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackServerGroupCachingAgent.groovy]: Oh! We don't support Kilo anymore, can you remove the Z with the timestamp? In a different PR of course :)

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/description/servergroup/ServerGroupParameters.groovy (hunk 1 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/description/servergroup/ServerGroupParameters.groovy]: user_data is the converted bytes. source_user_data is merely stored so we can retrieve and show back to the user.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackServerGroupCachingAgent.groovy (hunk 3 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackServerGroupCachingAgent.groovy]: go for it!

**Comment by derekolk** [comment]
LGTM

**Comment by lwander** [comment]
LGTM

**Pull request merged**